### PR TITLE
[#3649] feat(client): remove log4j impl from client-java-runtime

### PR DESCRIPTION
**Pull request by FANNG1**
### What changes were proposed in this pull request?
remove log4j impl from client-java-runtime

### Why are the changes needed?

Fix: #3649 

### Does this PR introduce _any_ user-facing change?
no

### How was this patch tested?
existing IT
local spark env

### CODE REVIEW

File: clients/client-java/src/main/java/com/datastrato/gravitino/client/RelationalCatalog.java [hunk 1]
```diff
@@ -239,12 +235,11 @@ public boolean purgeTable(NameIdentifier ident) throws UnsupportedOperationExcep
               ErrorHandlers.tableErrorHandler());
       resp.validate();
       return resp.dropped();
-
-    } catch (UnsupportedOperationException e) {
-      throw e;
     } catch (Exception e) {
-      LOG.warn("Failed to purge table {}", ident, e);
-      return false;
```
Comment (FANNG1) [inline clients/client-java/src/main/java/com/datastrato/gravitino/client/RelationalCatalog.java:247]: it's not correct to return false

File: core/src/main/java/com/datastrato/gravitino/Config.java [hunk 2]
```diff
@@ -1,11 +1,12 @@
 /*
- * Copyright 2023 Datastrato Pvt Ltd.
+ * Copyright 2024 Datastrato Pvt Ltd.
  * This software is licensed under the Apache License version 2.
  */
 package com.datastrato.gravitino;
```
Comment (FANNG1) [inline core/src/main/java/com/datastrato/gravitino/Config.java:5]: move `Config` from `common` module to `core` module, @jerryshao WDYT?

Comment (FANNG1) [reply core/src/main/java/com/datastrato/gravitino/Config.java:5]: If it's ok,  I will move `ConfigBuilder` and `ConfigEntry` and `ConfigConstants` too

File: clients/client-java/src/main/java/com/datastrato/gravitino/client/HTTPClient.java [hunk 3]
```diff
@@ -205,9 +201,6 @@ private void throwFailure(
           errorResponse =
               ((ErrorHandler) errorHandler).parseResponse(response.getCode(), responseBody, mapper);
         } else {
-          LOG.warn(
-              "Unknown error handler {}, response body won't be parsed",
-              errorHandler.getClass().getName());
```
Comment (FANNG1) [inline clients/client-java/src/main/java/com/datastrato/gravitino/client/HTTPClient.java:210]: error msg is put to errorResponse

File: core/src/main/java/com/datastrato/gravitino/Config.java (hunk 2 above)
Comment (FANNG1) [reply core/src/main/java/com/datastrato/gravitino/Config.java]: done

File: clients/client-java-runtime/build.gradle.kts [hunk 4]
```diff
@@ -11,7 +11,9 @@ plugins {
 }
 
 dependencies {
-  implementation(project(":clients:client-java"))
+  implementation(project(":clients:client-java")) {
+    exclude("org.apache.logging.log4j")
```
[the remark this answers is not in the document]
Comment (jerryshao) [reply clients/client-java-runtime/build.gradle.kts:15]: If we can make sure there's no log4j in dependency tree, then I think we can remove this.

[the remark this answers is not in the document]
Comment (FANNG1) [reply clients/client-java-runtime/build.gradle.kts]: removed and  add dependency check

File: clients/client-java-runtime/build.gradle.kts [hunk 5]
```diff
@@ -10,6 +10,14 @@ plugins {
...
 }
 
+configurations.all {
+  resolutionStrategy.eachDependency {
+    if (requested.group == "org.apache.logging.log4j") {
+      throw GradleException("Forbidden dependency 'org.apache.logging.log4j' found!")
+    }
+  }
+}
+
```
Comment (jerryshao) [inline clients/client-java-runtime/build.gradle.kts:20]: I don't think we have to do this explicit check here, if it is OK by manual check I think it is fine.

Comment (FANNG1) [reply clients/client-java-runtime/build.gradle.kts:20]: I prefer to keep this check in case other developers add log4j dependency unexpectedly.

**Comment by FANNG1** [comment]
@mchades  some changes about `purgeTable`, please help to review again

File: catalogs/catalog-jdbc-mysql/src/test/java/com/datastrato/gravitino/catalog/mysql/integration/test/CatalogMysqlIT.java [hunk 6]
```diff
@@ -1367,7 +1367,8 @@ void testNameSpec() {
 
     Assertions.assertTrue(catalog.asTableCatalog().dropTable(tableIdent));
     Assertions.assertFalse(catalog.asTableCatalog().tableExists(tableIdent));
-    Assertions.assertFalse(catalog.asTableCatalog().purgeTable(tableIdent));
+    Assertions.assertThrowsExactly(
```
Comment (FANNG1) [inline catalogs/catalog-jdbc-mysql/src/test/java/com/datastrato/gravitino/catalog/mysql/integration/test/CatalogMysqlIT.java:1370]: mysql doesn't support purge table

File: catalogs/catalog-hive/src/main/java/com/datastrato/gravitino/catalog/hive/HiveCatalogOperations.java [hunk 7]
```diff
@@ -1015,10 +1015,14 @@ public boolean dropTable(NameIdentifier tableIdent) {
...
-    } else {
-      return dropHiveTable(tableIdent, true, true);
+    try {
+      if (isExternalTable(tableIdent)) {
+        throw new UnsupportedOperationException("Can't purge a external hive table");
+      } else {
+        return dropHiveTable(tableIdent, true, true);
+      }
+    } catch (NoSuchTableException e) {
+      return false;
```
Comment (jerryshao) [inline catalogs/catalog-hive/src/main/java/com/datastrato/gravitino/catalog/hive/HiveCatalogOperations.java:1025]: @mchades can you please check this, is it the right fix, also I think we don't need to mix the changes here, if it is an issue we can use a separate PR to address it.

File: core/src/main/java/com/datastrato/gravitino/catalog/TableNormalizeDispatcher.java [hunk 8]
```diff
@@ -89,7 +89,9 @@ public boolean dropTable(NameIdentifier ident) {
 
   @Override
   public boolean purgeTable(NameIdentifier ident) throws UnsupportedOperationException {
-    return dispatcher.purgeTable(normalizeNameIdentifier(ident));
+    // The constraints of the name spec may be more strict than underlying catalog,
+    // and for compatibility reasons, we only apply case-sensitive capabilities here.
+    return dispatcher.purgeTable(applyCaseSensitive(ident, Capability.Scope.TABLE, dispatcher));
```
Comment (jerryshao) [inline core/src/main/java/com/datastrato/gravitino/catalog/TableNormalizeDispatcher.java:94]: Is this change necessary? @mchades can you please review it.

File: catalogs/catalog-hive/src/main/java/com/datastrato/gravitino/catalog/hive/HiveCatalogOperations.java (hunk 7 above)
Comment (mchades) [reply catalogs/catalog-hive/src/main/java/com/datastrato/gravitino/catalog/hive/HiveCatalogOperations.java:1025]: yeah, #3685 will track this, those changes should be reverted in this PR

**Comment by FANNG1** [comment]
@mchades @jerryshao I will remove the changes about `purgeTable` in latest commit.

**Comment by FANNG1** [comment]
> @mchades @jerryshao I will remove the changes about `purgeTable` in latest commit.

done, please help to review again @jerryshao

File: clients/client-java/src/main/java/com/datastrato/gravitino/client/HTTPClient.java [hunk 9]
```diff
@@ -222,7 +215,6 @@ private void throwFailure(
         // In such cases, we handle the situation by building an error response for the user.
         // Examples of such scenarios include network timeouts or load balancers returning default
         // 5xx responses.
-        LOG.error("Failed to parse an error response. Will create one instead.", e);
       }
```
Comment (mchades) [inline clients/client-java/src/main/java/com/datastrato/gravitino/client/HTTPClient.java:218]: do we need to remove this catch block?

File: clients/client-java/src/main/java/com/datastrato/gravitino/client/RelationalCatalog.java (hunk 1 above)
Comment (mchades) [reply clients/client-java/src/main/java/com/datastrato/gravitino/client/RelationalCatalog.java:247]: if so, could you please add a todo comment on int

**Review by jerryshao — APPROVED** [review]

File: clients/client-java/src/main/java/com/datastrato/gravitino/client/HTTPClient.java (hunk 9 above)
Comment (jerryshao) [reply clients/client-java/src/main/java/com/datastrato/gravitino/client/HTTPClient.java:218]: @FANNG1 can you please check this?

File: core/src/main/java/com/datastrato/gravitino/Config.java
Comment (mchades) [inline core/src/main/java/com/datastrato/gravitino/Config.java:1]: why move those classes?

Comment (FANNG1) [reply core/src/main/java/com/datastrato/gravitino/Config.java:1]: client doesn't need this,and contains some log hard to remove,so better to place in common

File: clients/client-java/src/main/java/com/datastrato/gravitino/client/HTTPClient.java (hunk 9 above)
Comment (FANNG1) [reply clients/client-java/src/main/java/com/datastrato/gravitino/client/HTTPClient.java:218]:
```java
   if (errorResponse == null) {
      errorResponse = buildRestErrorResponse(response);
    }
```
